Add exception to OperatorManifest yaml operations

lkolacek opened the pull request:
Closes CLOUDBLD-2939

Signed-off-by: Ladislav Kolacek <lkolacek@redhat.com>

lkolacek commented:
PTAL @pbortlov  @lipoja  @lcarva 

Looks like the right exception was not caught so the run did not fail. Do we want to also capture an error message from the exception and send it to the user? Unit test needed?

Who should I contact for reviews? Thanks!

MartinBasti commented on iib/workers/tasks/build.py:
```diff
@@ -1420,7 +1420,7 @@ def _apply_package_name_suffix(metadata_path, organization=None):
     with open(annotations_yaml_path, 'r') as f:
         try:
             annotations_yaml = yaml.load(f)
-        except ruamel.yaml.YAMLError:
```
I'm puzzled here, this is inheritance:

```
Exception --> YAMLError --> MarkedYAMLError --> ScannerError(MarkedYAMLError)
```
https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/scanner.py
https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/error.py

So scanner error should be already covered by `YAMLError`

MartinBasti replied:
https://github.com/release-engineering/iib/blob/master/iib/workers/tasks/build.py#L1501  This is the line that causes issues

lkolacek replied:
Thanks!

lcarva replied:
I don't think this is the right place. From the traceback:
```
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,624: DEBUG/ForkPoolWorker-1] The "package_name_suffix" configuration is not set for the organization certified-operators
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,719: INFO/ForkPoolWorker-1] Setting the state of request 19106 to "failed" with the reason "An unknown error occurred"
Oct 08 13:42:58 iib-main celery[2130]: [2020-10-08 13:42:58,719: INFO/ForkPoolWorker-1] Patching the request 19106 with {'state': 'failed', 'state_reason': 'An unknown error occurred'}
Oct 08 13:43:00 iib-main celery[2130]: [2020-10-08 13:43:00,512: ERROR/ForkPoolWorker-1] Task iib.workers.tasks.build.handle_regenerate_bundle_request[51f9e164-4f84-466c-8f51-176018d91c11] >
Oct 08 13:43:00 iib-main celery[2130]: Traceback (most recent call last):
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/celery/app/trace.py", line 385, in trace_task
Oct 08 13:43:00 iib-main celery[2130]:     R = retval = fun(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/celery/app/trace.py", line 650, in __protected_call__
Oct 08 13:43:00 iib-main celery[2130]:     return self.run(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/utils.py", line 263, in wrapper
Oct 08 13:43:00 iib-main celery[2130]:     return func(*args, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/build.py", line 1244, in handle_regenerate_bundle_request
Oct 08 13:43:00 iib-main celery[2130]:     manifests_path, metadata_path, organization, pinned_by_iib
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/iib/workers/tasks/build.py", line 1425, in _adjust_operator_bundle
Oct 08 13:43:00 iib-main celery[2130]:     operator_manifest = OperatorManifest.from_directory(manifests_path)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 555, in from_directory
Oct 08 13:43:00 iib-main celery[2130]:     operator_csvs = list(cls._get_csvs(yaml_files, **kwargs))
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 569, in _get_csvs
Oct 08 13:43:00 iib-main celery[2130]:     yield OperatorCSV.from_file(f, **kwargs)
Oct 08 13:43:00 iib-main celery[2130]:   File "/opt/iib-worker/virtualenv/lib/python3.6/site-packages/operator_manifest/operator.py", line 317, in from_file
Oct 08 13:43:00 iib-main celery[2130]:     data = yaml.load(f)
...
...
...
Oct 08 13:43:00 iib-main celery[2130]: ruamel.yaml.scanner.ScannerError: while scanning for the next token
Oct 08 13:43:00 iib-main celery[2130]: found character '\t' that cannot start any token
Oct 08 13:43:00 iib-main celery[2130]:   in "/tmp/iib-swyr59vk/manifests/cic-operator.v1.7.6.clusterserviceversion.yaml", line 195, column 9
```

That looks like this line: https://github.com/release-engineering/iib/blob/aec253346d6c51cb52293154f5008dc19048f079/iib/workers/tasks/build.py#L1501
The line number is off, but that's just because the code base has changed since October 8th.

It also matches my understanding of the issue. The error is seen when there are issues in the CSV file which defines the operator manifest, not the annotations YAML file.

I think it's important to catch the `ruamel.yaml.constructor.DuplicateKeyError` exception and provide a meaningful error message just because we see this so often in IIB. In reality, there could be many other errors, not all related to invalid YAML. 

The problem with not catching errors is that the traceback is logged on the server side and the user has no idea why the IIB build failed. There's something simple we can do this to save ourselves some headache in the future. For all IIB celery tasks, this error callback is installed, https://github.com/release-engineering/iib/blob/master/iib/workers/tasks/general.py#L14

We can leverage this to ensure any error is logged properly and visible to the user. To do this, you can add the `request_logger` decorator to that task, similarly to how it's done to the other celery tasks in build.py, for example. Then, a simple call to `log.error(msg, exc_info=exc)` within `failed_request_callback` will make sure the error is propagated back to the user.

It would be great to do both of those things:
1. Provide a meaningful error for duplicated keys in the operator manifest, and
2. Always log errors in the build specific log file

MartinBasti commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,13 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
+    try:
+        operator_manifest = OperatorManifest.from_directory(manifests_path)
+    except ruamel.yaml.YAMLError:
+        error = 'metadata/annotations/yaml is not valid YAML'
```
:eyes:

Are you sure?

Would be possible to don't repeat the same code, maybe using context manager/decorator/whatever??

lkolacek replied:
changed

MartinBasti commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,13 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
+    except ruamel.yaml.YAMLError:
+        error = 'metadata/annotations/yaml is not valid YAML'
+        log.exception(error)
+        raise IIBError(error)
```
Instead of using generic error message, wouldn't be better to use concrete error message returned by  ruamel?

Like `f"not a valid yaml file: {exc}"`

lkolacek replied:
Changed

MartinBasti commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,13 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
-    operator_manifest = OperatorManifest.from_directory(manifests_path)
+    try:
+        operator_manifest = OperatorManifest.from_directory(manifests_path)
+    except ruamel.yaml.YAMLError:
```
Are you sure that just this error can be raised from `OperatorManifest.from_directory`

https://github.com/containerbuildsystem/operator-manifest/blob/master/operator_manifest/operator.py#L579

Shouldn't we just catch all the exception and properly log and return them?

lkolacek replied:
@MartinBasti  @yashvardhannanavati  I added (RuntimeError, ValueError) , let me know if you want to catch also general exception. Unit tests added.

yashvardhannanavati commented:
@lkolacek could you please update unit tests for it?

MartinBasti commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,17 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
+        log.exception(error)
+        raise IIBError(error)
+    except (RuntimeError, ValueError) as e:
+        error = f'OperatorManifest creation failed: {e}'
```
I'd say loading or parsing or something. Operator files already exist, it may puzzle users

lkolacek replied:
Changed to loading  then

MartinBasti commented on tests/test_workers/test_tasks/test_build.py:
```diff
@@ -1490,6 +1490,44 @@ def test_adjust_operator_bundle_invalid_related_images(mock_apns, tmpdir):
+    )
+
+    with pytest.raises(IIBError, match=expected):
+        build._adjust_operator_bundle(str('/foo'), str('/bar'))
```
:eyes: 

why do you need retype string to string?

lkolacek commented:
I would like to ask for final review, Travis CI was fixed with adding missing sha
@MartinBasti  @yashvardhannanavati

MartinBasti commented:
https://media.giphy.com/media/ALalVMOVR8Qw/giphy.gif

2 unrelated fixes in single commit :'(

pbortlov commented on requirements-test.txt:
```diff
@@ -183,6 +183,7 @@ mako==1.1.1 \
     --hash=sha256:2984a6733e1d472796ceef37ad48c26f4a984bb18119bb2dbc37a44d8f6e75a4 \
     # via alembic
 markupsafe==1.1.1 \
+    --hash=sha256:a6a744282b7718a2a62d2ed9d993cad6f5f585605ad352c11de459f4108df0a1 \
```
I would prefer if the changes in requirements are at least in the different commit.

pbortlov commented on requirements.txt:
```diff
@@ -124,6 +124,7 @@ mako==1.1.1 \
     --hash=sha256:2984a6733e1d472796ceef37ad48c26f4a984bb18119bb2dbc37a44d8f6e75a4 \
     # via alembic
 markupsafe==1.1.1 \
+    --hash=sha256:a6a744282b7718a2a62d2ed9d993cad6f5f585605ad352c11de459f4108df0a1 \
```
I would prefer if the changes in requirements are at least in the different commit.

lkolacek commented:
@pbortlov  @MartinBasti  Ok I just found out that there is a procedure how to update sha in requirements.
* https://github.com/release-engineering/iib#dependency-management

I'll deal with that in different commit. Thanks for info @yashvardhannanavati

MartinBasti reviewed (approved):
:rocket:

lcarva commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,17 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
+        error = f'Not a valid yaml file: {e}'
+        log.exception(error)
+        raise IIBError(error)
+    except (RuntimeError, ValueError) as e:
```
@lkolacek, I came back from PTO and forgot to refresh the tab. I did a silly thing an commented on an outdated diff. Could you please have a look at the second part of my [comment](https://github.com/release-engineering/iib/pull/209#discussion_r570972525) where I suggest logging from the error callback? 

We might not need to catch `RuntimeError` and `ValueError` here.

lkolacek replied:
I added `@request_logger` as you requested. 

Catching also other exceptions except `ruamel.yaml.YAMLError` suggested Martin and I agree with him. Even though there is only a slight chance that this problem will ever occur.

yashvardhannanavati commented:
@lkolacek Could you please rebase your PR? I updated the markupsafe dep in a different PR. Sorry for the trouble

lcarva commented on iib/workers/tasks/build.py:
```diff
@@ -1498,7 +1498,13 @@ def _adjust_operator_bundle(manifests_path, metadata_path, organization=None, pi
+    try:
+        operator_manifest = OperatorManifest.from_directory(manifests_path)
+    except ruamel.yaml.YAMLError as e:
+        error = f'Not a valid yaml file: {e}'
```
Can we be more specific? How about: "The Operator Manifest is not in a valid YAML format: {e}"

lcarva commented:
> @lkolacek Could you please rebase your PR? I updated the markupsafe dep in a different PR. Sorry for the trouble

@lkolacek, maybe we need to drop that first commit?

lkolacek commented:
> > @lkolacek Could you please rebase your PR? I updated the markupsafe dep in a different PR. Sorry for the trouble
> 
> @lkolacek, maybe we need to drop that first commit?

I apologize, I missed this comment. I rebased before I left and saw no collision because we added sha to 2 different places. My commit was removed.
